unixfs/examples/cat.rs

```diff
@@ -0,0 +1,161 @@
+use cid::Cid;
+use ipfs_unixfs::file::{visit::IdleFileVisit, FileReadFailed};
+use std::convert::TryFrom;
+use std::fmt;
+use std::io::{Error as IoError, Read, Write};
+use std::path::PathBuf;
+
+fn main() {
+    let cid = match std::env::args().skip(1).next().map(|s| Cid::try_from(s)) {
+        Some(Ok(cid)) => cid,
+        Some(Err(e)) => {
+            eprintln!("Invalid cid given as argument: {}", e);
+            std::process::exit(1);
+        }
+        None => {
+            eprintln!("USAGE: {} CID", std::env::args().next().unwrap());
+            eprintln!();
+            eprintln!(
+                "Will walk the unixfs file pointed out by the CID from default go-ipfs 0.5 \
+                configuration flatfs blockstore and write all content to stdout."
+            );
+            std::process::exit(0);
+        }
+    };
+
+    let ipfs_path = match std::env::var("IPFS_PATH") {
+        Ok(s) => s,
+        Err(e) => {
+            eprintln!("IPFS_PATH is not set or could not be read: {}", e);
+            std::process::exit(1);
+        }
+    };
+
+    let mut blocks = PathBuf::from(ipfs_path);
+    blocks.push("blocks");
+
+    let blockstore = ShardedBlockStore { root: blocks };
+
+    match walk(blockstore, &cid) {
+        Ok((read, content)) => {
+            eprintln!("Content bytes: {}", content);
+            eprintln!("Total bytes:   {}", read);
+        }
+        Err(Error::OpeningFailed(e)) => {
+            eprintln!("{}", e);
+            eprintln!();
+            eprintln!("This is likely caused by either:");
+            eprintln!(" - ipfs does not have the block");
+            eprintln!(" - ipfs is configured to use non-flatfs storage");
+            eprintln!(" - ipfs is configured to use flatfs with different sharding");
+            std::process::exit(1);
+        }
+        Err(e) => {
+            eprintln!("Failed to walk the merkle tree: {}", e);
+            std::process::exit(1);
+        }
+    }
+}
+
+fn walk(blocks: ShardedBlockStore, start: &Cid) -> Result<(u64, u64), Error> {
+    let stdout = std::io::stdout();
+    let mut stdout = stdout.lock();
+
+    let mut read_bytes = 0;
+    let mut content_bytes = 0;
+
+    // The blockstore specific way of
+    let mut buf = Vec::new();
+    read_bytes += blocks.as_file(&start.to_bytes())?.read_to_end(&mut buf)? as u64;
```

Comment:
just to be sure: we don't expect `ShardedBlockStore` to be big enough to cause trouble in memory-constrained scenarios? otherwise we might prefer to use a `BufReader` approach instead of `read_to_end`.

Reply:
Note, this is the example code. I think adding BufReader here would only increase the number of copies, but reading in smaller parts might be good.

In the bigger picture the blocks are currently only loaded at once. Streaming will be an interesting to experiment with one day, but at the moment blocks are expected to be small enough (I can never remember the limit) so that they can be always fully loaded into memory.

For streaming support my `minipb` would be missing some bug fixes and an ability to read the length delimited bytes in slices. The API of this crate would need to be adapted to return multiple contents when there are such. I don't think it would be nearly as usable as this is :)